### PR TITLE
feat(protonvpn): Added ProtonVPN feature selection

### LINKED ISSUE #1582: Feature request: Protonvpn filtering by features
### What's the feature 🧐

Each server JSON object from their API has a feature number corresponding to a feature detailed in https://github.com/ProtonVPN/protonvpn-nm-lib/blob/31d5f99fbc89274e4e977a11e7432c0eab5a3ef8/protonvpn_nm_lib/enums.py#L44-L49

Gluetun should allow to filter servers by feature

### Extra information and references

_No response_

### PR DESCRIPTION
**Pull request by kvanzuijlen**
I know it's not completely done yet (missing tests, missing todo's for moving feature-related stuff to the new Features struct, docs, etc) but I wanted to open a PR to see if you agree with my chosen direction.

Closes #1582, partially #1103

### CODE REVIEW

File: internal/configuration/settings/serverselection.go [hunk 1]
```diff
@@ -54,21 +54,33 @@ type ServerSelection struct { //nolint:maligned
 	// be filtered. This is used with VPN Secure.
 	// TODO extend to providers using FreeOnly.
 	PremiumOnly *bool `json:"premium_only"`
+	// OpenVPN contains settings to select OpenVPN servers
+	// and the final connection.
+	OpenVPN OpenVPNSelection `json:"openvpn"`
+	// Wireguard contains settings to select Wireguard servers
+	// and the final connection.
+	Wireguard WireguardSelection `json:"wireguard"`
```
Comment (qdm12) [inline internal/configuration/settings/serverselection.go:62]: nit Please move back to the bottom of the struct to reduce deltas / git blame 😉 Thanks!

File: internal/configuration/settings/serverselection.go [hunk 2]
```diff
@@ -107,6 +123,11 @@ func (ss *ServerSelection) validate(vpnServiceProvider string,
 		return fmt.Errorf("for VPN service provider %s: %w", vpnServiceProvider, err)
 	}
 
+	err = validateFeatureFilters(*ss, vpnServiceProvider)
+	if err != nil {
+		return fmt.Errorf("for VPN service provider %s: %w", vpnServiceProvider, err)
+	}
+
 	if *ss.OwnedOnly &&
 		vpnServiceProvider != providers.Mullvad {
```
Comment (qdm12) [inline internal/configuration/settings/serverselection.go:132]: Why not incorporate OwnedOnly etc. (below) within validateFeatureFilters?

File: internal/configuration/settings/serverselection.go [hunk 3]
```diff
@@ -107,6 +123,11 @@ func (ss *ServerSelection) validate(vpnServiceProvider string,
 		return fmt.Errorf("for VPN service provider %s: %w", vpnServiceProvider, err)
 	}
 
+	err = validateFeatureFilters(*ss, vpnServiceProvider)
+	if err != nil {
+		return fmt.Errorf("for VPN service provider %s: %w", vpnServiceProvider, err)
```
Comment (qdm12) [inline internal/configuration/settings/serverselection.go:128]:
```suggestion
		return err
```

since the provider name is already in the returned error

File: internal/configuration/settings/serverselection.go [hunk 4]
```diff
@@ -54,21 +54,33 @@ type ServerSelection struct { //nolint:maligned
...
+	SecureCoreOnly *bool `json:"secure_core_only"`
+	// TorOnly is true if VPN servers without tor should
+	// be filtered. This is used with ProtonVPN.
+	TorOnly *bool `json:"tor_only"`
+	// P2POnly is true if VPN servers not for p2p should
+	// be filtered. This is used with ProtonVPN.
+	P2POnly *bool `json:"p2p_only"`
+	// IPv6Only is true if VPN servers without IPv6 support should
+	// be filtered. This is used with ProtonVPN.
+	IPv6Only *bool `json:"ipv6_only"`
```
Comment (qdm12) [inline internal/configuration/settings/serverselection.go:83]: What does this filter do? 🤔 
Because AFAIK there are only IPv4 addresses to reach Protonvpn servers.

File: .gitignore [hunk 5]
```diff
@@ -1 +1,2 @@
 scratch.txt
+.idea/
```
Comment (qdm12) [inline .gitignore:2]: nit remove before merge 😉 Or we can do a separate PR for it

```suggestion
```

File: internal/models/server.go [hunk 6]
```diff
@@ -29,11 +29,20 @@ type Server struct {
...
 	IPs         []netip.Addr `json:"ips,omitempty"`
+	Features    *Features    `json:"features,omitempty"`
+}
+
+type Features struct {
+	SecureCore bool `json:"secure_core,omitempty"`
+	Tor        bool `json:"tor,omitempty"`
+	P2P        bool `json:"p2p,omitempty"`
+	Streaming  bool `json:"stream,omitempty"`
+	IPv6       bool `json:"ipv6,omitempty"`
```
Comment (qdm12) [inline internal/models/server.go:45]: What's the advantage over having them directly as fields of `Server` as they were?
I see it makes it a bit tidier indeed, but to keep it simple/retro-compatible/no-change-needed, I would keep the same structure as before. You can however add a `// TODO v4 ...` comment describing it would be nice to have them in a Features struct though.

File: internal/provider/protonvpn/updater/api.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ type logicalServer struct {
 	Region      *string          `json:"Region"`
 	City        *string          `json:"City"`
 	Servers     []physicalServer `json:"Servers"`
+	Features    int              `json:"Features"`
```
Comment (qdm12) [inline internal/provider/protonvpn/updater/api.go:26]: nit I guess these are positive only numbers, so let's use uint

```suggestion
	Features    uint              `json:"Features"`
```

Or maybe event `uint16` I think? 🤔

File: internal/provider/protonvpn/updater/iptoserver.go [hunk 8]
```diff
@@ -10,7 +10,7 @@ import (
 type ipToServer map[string]models.Server
 
 func (its ipToServer) add(country, region, city, name, hostname string,
-	free bool, entryIP netip.Addr) {
+	free bool, entryIP netip.Addr, features *models.Features) {
```
Comment (qdm12) [inline internal/provider/protonvpn/updater/iptoserver.go:13]: nit just return the whole bunch of features after `free bool`, or define an unexported `features` struct in this package, up to you 😉 (and then add the `free bool` to it too.

File: internal/provider/utils/filtering.go [hunk 9]
```diff
@@ -45,7 +45,7 @@ func filterServer(server models.Server,
 		return true
 	}
 
-	if *selection.StreamOnly && !server.Stream {
+	if *selection.StreamOnly && (!server.Stream || !server.Features.Streaming) {
```
Comment (qdm12) [inline internal/provider/utils/filtering.go:48]: We should have a single `Stream` field, so probably just set the value of `Features.Streaming` into `Stream`.

File: internal/storage/filter.go [hunk 10]
```diff
@@ -67,7 +67,7 @@ func filterServer(server models.Server,
 		return true
 	}
 
-	if *selection.StreamOnly && !server.Stream {
+	if *selection.StreamOnly && (!server.Stream || !server.Features.Streaming) {
```
Comment (qdm12) [inline internal/storage/filter.go:70]: same comment here, we need to merge both fields

**Review by qdm12 — COMMENTED** [review]
Nice it's looking pretty good, just a few minor comments 😉 👍

File: internal/configuration/settings/serverselection.go (hunk 1 above)
Comment (kvanzuijlen) [reply internal/configuration/settings/serverselection.go:62]: Will do!

File: internal/configuration/settings/serverselection.go (hunk 2 above)
Comment (kvanzuijlen) [reply internal/configuration/settings/serverselection.go:132]: Not sure if this was a feature or a "subscription tier" filter. If it is a feature then it makes sense to move it as well!

File: internal/models/server.go (hunk 6 above)
Comment (kvanzuijlen) [reply internal/models/server.go:45]: it was mostly not to add too many parameters to the [iptoserver.go add](https://github.com/qdm12/gluetun/pull/2182/files#diff-24ccd1d540dfbe5e9a4e94c5fb8e12ac3ffa3bc0a425a2cd0d70557c51b2eefdR12-R13) function. I'm also fine with just making it part of the server struct though.

File: internal/provider/protonvpn/updater/api.go (hunk 7 above)
Comment (kvanzuijlen) [reply internal/provider/protonvpn/updater/api.go:26]: Not familiar enough with Go 😅. Both uint16 and uint sound fine to me, with uint having the advantage of not having to be changed if a feature is added and uint16 not reserving unneeded memory space (if I understand correctly). Let me know which of the 2 has your preference!

File: internal/provider/protonvpn/updater/iptoserver.go (hunk 8 above)
Comment (kvanzuijlen) [reply internal/provider/protonvpn/updater/iptoserver.go:13]: Will wait with this to see what you decide about the features struct in models/server.go

File: internal/provider/utils/filtering.go (hunk 9 above)
Comment (kvanzuijlen) [reply internal/provider/utils/filtering.go:48]: Will do! (optionally with a TODO for v4, depending on what you decide about the features struct)

File: internal/configuration/settings/serverselection.go (hunk 4 above)
Comment (kvanzuijlen) [reply internal/configuration/settings/serverselection.go:83]: Not sure! It's indeed not a supported feature by any current server, but I just wanted to implement all features listed in [protonvpn-nm-lib](https://github.com/ProtonVPN/protonvpn-nm-lib/blob/master/protonvpn_nm_lib/enums.py#L44-L49)

File: internal/provider/protonvpn/updater/api.go (hunk 7 above)
Comment (qdm12) [reply internal/provider/protonvpn/updater/api.go:26]: `uint` has the advantage to reject negative inputs and have compilation safety against negative numbers.... just in case ™️ 

`uint16` restricts it further (from 64 or 32 bits to 16 bits on all platforms), and perhaps highlight more it's used as a bit array and not just a number. Let's try uint16 😉

File: internal/models/server.go (hunk 6 above)
Comment (qdm12) [reply internal/models/server.go:45]: Let's have these new fields directly in the Server struct, and add an unexported `features` (no first capital letter for unexported types) struct in the updater package to keep the code as clean as it is 😉

File: internal/provider/protonvpn/updater/iptoserver.go (hunk 8 above)
Comment (kvanzuijlen) [reply internal/provider/protonvpn/updater/iptoserver.go]: The features are defined at logical server level, "free" at physical server level, didn't feel right to put it both in the same struct, but will do if you prefer it.

**Comment by kvanzuijlen** [comment]
Hi,

I applied (most of) your suggestions from the code review. I also added minimal unit tests (basically added to all the _test files) for all files changed in this PR. I also built my own image and started using it already. It seems to work, as far as I can tell.

**Comment by kvanzuijlen** [comment]
@qdm12 I've been running this code for a couple of days now myself (https://github.com/users/kvanzuijlen/packages/container/package/gluetun) and as far as I can tell it now fully works :).

File: internal/configuration/settings/serverselection.go (hunk 2 above)
Comment (qdm12) [reply internal/configuration/settings/serverselection.go]: Good point indeed, I've separated the validation in `validateSubscriptionTierFilters` and `validateFeatureFilters`

File: internal/configuration/settings/serverselection.go (hunk 4 above)
Comment (qdm12) [reply internal/configuration/settings/serverselection.go:77]: Removed due to the strange filter option name `IPV6_ONLY` and because it's unused. We can always re-add it later if needed.

File: internal/provider/protonvpn/updater/iptoserver.go (hunk 8 above)
Comment (qdm12) [reply internal/provider/protonvpn/updater/iptoserver.go]: This makes sense, especially since free is more of a subscription tier filter as you mentioned 😉

**Comment by qdm12** [comment]
1. I rebased your branch on master
2. Extracting features was actually semi-functional... but fear not, it's Go's `switch` which behaves strangely compared to other programming languages. When encountering a true case in the switch, it gets out of the switch and doesn't try other cases. That means if the features bits match one feature, it won't find other features. This is fixed by assigning the features struct fields directly without a switch:

    ```go
    features := features{
    	secureCore: featuresBits&1 != 0,
    	tor:        featuresBits&2 != 0,
    	p2p:        featuresBits&4 != 0,
    	stream:     featuresBits&8 != 0,
    	// ipv6: featuresBits&16 != 0, - unused.
    }
    ```

   FYI this could also been fixed using the `fallthrough` keyword at the end of each case block in the switch 😉 
3. IPv6 only filter got removed. It's a bit of a confusing option `IPV6_ONLY` and also does not seem to be used.
4. Re-updated servers data
5. Added new environment variable options to Dockerfile

    ```Dockerfile
    SECURE_CORE_ONLY= \
    TOR_ONLY= \
    P2P_ONLY= \
    ```

**Review by qdm12 — APPROVED** [review]

**Comment by qdm12** [comment]
@kvanzuijlen are you aware of a difference between 'p2p' and 'port forwarding' in terms of feature for Protonvpn? I'm looking at https://github.com/qdm12/gluetun/issues/1103#issuecomment-2255404791 and thinking of removing P2P_ONLY in favor of PORT_FORWARD_ONLY. Also are **all** p2p servers able to port forward?
